rs/ledger_suite/icrc1/ledger/src/main.rs

```diff
@@ -237,21 +237,19 @@ fn post_upgrade(args: Option<LedgerArgument>) {
 
     PRE_UPGRADE_INSTRUCTIONS_CONSUMED.with(|n| *n.borrow_mut() = pre_upgrade_instructions_consumed);
 
-    if upgrade_from_version < LEDGER_VERSION {
-        if upgrade_from_version == 0 {
-            log_message("Upgrading from version 0 which does not use stable structures, clearing stable allowance data.");
-            clear_stable_allowance_data();
-        }
+    if upgrade_from_version == 0 {
         set_ledger_state(LedgerState::Migrating(LedgerField::Allowances));
+        log_message("Upgrading from version 0 which does not use stable structures, clearing stable allowance data.");
+        clear_stable_allowance_data();
         Access::with_ledger_mut(|ledger| {
             ledger.clear_arrivals();
         });
+    }
+    if !is_ready() {
```

Comment:
Why a separate check instead of continuing the previous if-branch? In any case, this means that if the migration fails, one needs to downgrade first: it is not possible to retrigger migration by upgrading to the same version, right?

Reply:
In V4 I will add a check above `if upgrade_from_version == 0 {`: `if upgrade_from_version < 2 { state= migrate_balances}`. It is possible to retrigger migration by upgrading to the same version, `test_incomplete_migration_to_current` tests that.

Reply:
If migration is not complete, `pre_upgrade` doesn't do anything, so the next upgrade sees the ledger as if it was being upgraded from V2 again - doesn't matter if the next upgrade is to V2 or again to V3.